src/handlers/proxy.js

```diff
@@ -11,42 +12,55 @@ const {
 } = http2.proxies;
 
 /**
- *
- * @param {string|undefined} rawProxyUrl
- * @returns {import('got/dist/source').HandlerFunction} - got handler.
+ * @param {object} options
+ * @param {function} next
+ * @returns {import('got').GotReturn}
  */
-exports.createProxyHandler = (rawProxyUrl) => {
-    if (rawProxyUrl) {
-        return (options, next) => {
-            options.agent = getAgent(rawProxyUrl, options.http2);
+async function proxyHandler(options, next) {
+    const { context: { proxyUrl } } = options;
 
-            return next(options);
-        };
+    if (proxyUrl) {
+        const parsedProxy = new URL(proxyUrl);
+
+        validateProxyProtocol(parsedProxy.protocol);
+        options.agent = await getAgent(parsedProxy, options.https.rejectUnauthorized);
     }
 
-    return (options, next) => next(options);
-};
+    return next(options);
+}
+
+/**
+ * @param {string} protocol
+ */
+function validateProxyProtocol(protocol) {
+    const isSupported = protocol === 'http:' || protocol === 'https:';
+
+    if (!isSupported) {
+        throw new Error(`Invalid proxy protocol "${protocol}"`);
```

Comment:
The fact that we don't support anything else does not mean that various protocols are "invalid". I would say something like "Proxy URL protocol {protocol} is not supported. Please use HTTP or HTTPS."